builder/builder.go

```diff
@@ -0,0 +1,694 @@
+	transactions []bellatrix.Transaction,
+) *builderApi.VersionedSubmitBlindedBlockResponse {
+	return &builderApi.VersionedSubmitBlindedBlockResponse{
+		Version: spec.DataVersionBellatrix,
+		Bellatrix: &bellatrix.ExecutionPayload{
+			ParentHash:    [32]byte(payload.ParentHash),
+			FeeRecipient:  [20]byte(payload.FeeRecipient),
+			StateRoot:     [32]byte(payload.StateRoot),
+			ReceiptsRoot:  [32]byte(payload.ReceiptsRoot),
+			LogsBloom:     types.BytesToBloom(payload.LogsBloom),
+			PrevRandao:    [32]byte(payload.Random),
+			BlockNumber:   payload.Number,
+			GasLimit:      payload.GasLimit,
+			GasUsed:       payload.GasUsed,
+			Timestamp:     payload.Timestamp,
+			ExtraData:     payload.ExtraData,
+			BaseFeePerGas: baseFeePerGas,
+			BlockHash:     [32]byte(payload.BlockHash),
+			Transactions:  transactions,
+		},
+	}
+}
+
+func getCapellaPayload(
+	payload *engine.ExecutableData,
+	baseFeePerGas [32]byte,
+	transactions []bellatrix.Transaction,
+	withdrawals []*capella.Withdrawal,
+) *builderApi.VersionedSubmitBlindedBlockResponse {
+	return &builderApi.VersionedSubmitBlindedBlockResponse{
+		Version: spec.DataVersionCapella,
+		Capella: &capella.ExecutionPayload{
+			ParentHash:    [32]byte(payload.ParentHash),
+			FeeRecipient:  [20]byte(payload.FeeRecipient),
+			StateRoot:     [32]byte(payload.StateRoot),
+			ReceiptsRoot:  [32]byte(payload.ReceiptsRoot),
+			LogsBloom:     types.BytesToBloom(payload.LogsBloom),
+			PrevRandao:    [32]byte(payload.Random),
+			BlockNumber:   payload.Number,
+			GasLimit:      payload.GasLimit,
+			GasUsed:       payload.GasUsed,
+			Timestamp:     payload.Timestamp,
+			ExtraData:     payload.ExtraData,
+			BaseFeePerGas: baseFeePerGas,
+			BlockHash:     [32]byte(payload.BlockHash),
+			Transactions:  transactions,
+			Withdrawals:   withdrawals,
+		},
+	}
+}
+
+func getBlobsBundle(blobsBundle *engine.BlobsBundleV1) *builderApiDeneb.BlobsBundle {
+	commitments := make([]deneb.KZGCommitment, len(blobsBundle.Commitments))
+	proofs := make([]deneb.KZGProof, len(blobsBundle.Proofs))
+	blobs := make([]deneb.Blob, len(blobsBundle.Blobs))
+
+	// we assume the lengths for blobs bundle is validated beforehand to be the same
+	for i := range blobsBundle.Blobs {
+		var commitment deneb.KZGCommitment
+		copy(commitment[:], blobsBundle.Commitments[i][:])
+		commitments[i] = commitment
+
+		var proof deneb.KZGProof
+		copy(proof[:], blobsBundle.Proofs[i][:])
+		proofs[i] = proof
+
+		var blob deneb.Blob
+		copy(blob[:], blobsBundle.Blobs[i][:])
+		blobs[i] = blob
+	}
+	return &builderApiDeneb.BlobsBundle{
+		Commitments: commitments,
+		Proofs:      proofs,
+		Blobs:       blobs,
+	}
+}
+
+func getDenebPayload(
+	payload *engine.ExecutableData,
+	baseFeePerGas *uint256.Int,
+	transactions []bellatrix.Transaction,
+	withdrawals []*capella.Withdrawal,
+	blobsBundle *engine.BlobsBundleV1,
+) *builderApi.VersionedSubmitBlindedBlockResponse {
+
+	blobGasUsed := uint64(0)
+	if payload.BlobGasUsed != nil {
```

Comment:
when is this not set? should be set in [worker.go](https://github.com/flashbots/op-geth/pull/2/files#diff-689874b30f6f905ce6c47cdefeddcf052b467a7936a981f4b6cf38939e10d924R1305)

Reply:
fa67ef456cff244ae196929c4116a9899ba90766